server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -1110,35 +1109,31 @@ private void assertResponsesAreCorrect(BulkItemResponse[] bulkResponses, BulkIte
             }
         }
 
-        synchronized void markItemAsDropped(int slot) {
-            IndexRequest indexRequest = getIndexWriteRequest(bulkRequest.requests().get(slot));
-            failedSlots.set(slot);
-            final String id = indexRequest.id() == null ? DROPPED_ITEM_WITH_AUTO_GENERATED_ID : indexRequest.id();
-            itemResponses.add(
-                BulkItemResponse.success(
-                    slot,
-                    indexRequest.opType(),
-                    new UpdateResponse(
-                        new ShardId(indexRequest.index(), IndexMetadata.INDEX_UUID_NA_VALUE, 0),
-                        id,
-                        SequenceNumbers.UNASSIGNED_SEQ_NO,
-                        SequenceNumbers.UNASSIGNED_PRIMARY_TERM,
-                        indexRequest.version(),
-                        DocWriteResponse.Result.NOOP
-                    )
-                )
-            );
-        }
-
         synchronized void markItemAsFailed(int slot, Exception e) {
-            IndexRequest indexRequest = getIndexWriteRequest(bulkRequest.requests().get(slot));
+            final DocWriteRequest<?> docWriteRequest = bulkRequest.requests().get(slot);
+            final String id = Objects.requireNonNullElse(docWriteRequest.id(), DROPPED_OR_FAILED_ITEM_WITH_AUTO_GENERATED_ID);
             // We hit a error during preprocessing a request, so we:
             // 1) Remember the request item slot from the bulk, so that when we're done processing all requests we know what failed
             // 2) Add a bulk item failure for this request
             // 3) Continue with the next request in the bulk.
             failedSlots.set(slot);
-            BulkItemResponse.Failure failure = new BulkItemResponse.Failure(indexRequest.index(), indexRequest.id(), e);
-            itemResponses.add(BulkItemResponse.failure(slot, indexRequest.opType(), failure));
+            BulkItemResponse.Failure failure = new BulkItemResponse.Failure(docWriteRequest.index(), id, e);
+            itemResponses.add(BulkItemResponse.failure(slot, docWriteRequest.opType(), failure));
+        }
+
+        synchronized void markItemAsDropped(int slot) {
+            final DocWriteRequest<?> docWriteRequest = bulkRequest.requests().get(slot);
+            final String id = Objects.requireNonNullElse(docWriteRequest.id(), DROPPED_OR_FAILED_ITEM_WITH_AUTO_GENERATED_ID);
+            failedSlots.set(slot);
+            UpdateResponse dropped = new UpdateResponse(
```

Comment:
Why is this an UpdateResponse? I see that it was previously, but it leads to some weird things, like getting an UpdateResponse from an IndexRequest (which you also had to account for in this PR).

Reply:
I don't think IndexResponse lets us express anything except `CREATED` or `UPDATED`:

```
    private static Result assertCreatedOrUpdated(Result result) {
        assert result == Result.CREATED || result == Result.UPDATED;
        return result;
    }
```

But yeah, it is inconvenient (and also ancient!).